### ISSUE TITLE
Bug: parsing a function with argument throws an error when `range` parse option is not set

### ISSUE THREAD

**Issue by ericmorand** [typescript-eslint/typescript-eslint#9116]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

parser

### Repro Code

```TypeScript
import {parse} from "@typescript-eslint/parser";

parse(`function foo(bar) {
    bar;
}`);
```

### Expected Result

I expect that the `parse` call doesn't throw an error.

### Actual Result

The `parse` call throws the error

```
TypeError: Cannot read properties of undefined (reading '0')
    at FunctionScope.isValidResolution ([...]/node_modules/@typescript-eslint/scope-manager/dist/scope/FunctionScope.js:27:49)
```

### Additional Info

This is because the method `FunctionScope.isValidResolution` acts as if the range property of the body node was guaranteed to be present, while it is not. It is guaranteed to be present only if the `range` option is passed as `true` to the `parse` method.

I'd like to add that it is a global issue: the TypeScript declaration of the Node type acts as if the `range` property is not optional, which is wrong.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/parser`            | `7.9.0` |

**Comment by kirkwaiblinger** [comment]
Also reported on SO: https://stackoverflow.com/questions/78342447/im-using-typescript-eslint-parser-to-parse-a-typescript-function-how-can-i-ge#78434221

**Comment by bradzacher** [comment]
Duplicate of https://github.com/typescript-eslint/typescript-eslint/issues/8415

**Comment by ericmorand** [comment]
I did and I found the issue. It was closed but it is still happening. Please consider reopening.

**Comment by bradzacher** [comment]
Yes, but did you read the response in the issue which explains how to use it correctly and why it works this way...? 

https://github.com/typescript-eslint/typescript-eslint/issues/8415#issuecomment-1935227631

**Comment by ericmorand** [comment]
I did not, you are right. I should have.

Out of curiosity, why not just fixing the issue - i.e. removing the range option and update the code so that the range is always emitted - since this is the only way to use the package anyway? Why leaving the option to the end user?

**Comment by bradzacher** [comment]
As mentioned in the linked comment

> `parser` is not intended to be consumed by users - it's intended for use as an eslint parser.
> ...
> if you just want the AST - use our `typescript-estree` package instead (`parser` is just a wrapper on top of that package). This package is intended for user consumption as a parser.
